Connect: Envoy sidecars enter state where they fail to receive xds updates from Consul

### Overview of the Issue

We are experiencing issues with reliability in Consul Connect, specifically around connection and CDS issues, where for some reason, intermittently, Envoy sidecars (deployed with the `sidecar_service` stanza in Nomad) enter an unrecoverable state where they no longer receive cluster updates from Consul.

You can see some of this in #10213, and I've provided some detail there I'm copying here.

Essentially the only fix for this currently is to restart either Envoy sidecar itself, or the Nomad allocation (which then cycles the sidecar).

#### Reproduction Steps

Unable to do so at this time, however, it consistently happens at random times throughout the day.

### Consul info for both Client and Server

<details>
  <summary>Client info</summary>

```
$ consul info
agent:
	check_monitors = 1
	check_ttls = 39
	checks = 159
	services = 123
build:
	version = 1.10.0
consul:
	acl = disabled
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 12
	goroutines = 2096
	max_procs = 12
	os = linux
	version = go1.16.5
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 106
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 67643
	members = 230
	query_queue = 0
	query_time = 4
```

</details>

<details>
  <summary>Server info</summary>

```
~# consul info
agent:
	check_monitors = 1
	check_ttls = 0
	checks = 5
	services = 4
build:
	version = 1.10.0
consul:
	acl = disabled
	bootstrap = false
	known_datacenters = 4
	leader = false
	leader_addr = 10.133.0.146:8300
	server = true
raft:
	applied_index = 322772439
	commit_index = 322772439
	fsm_pending = 0
	last_contact = 32.532339ms
	last_log_index = 322772440
	last_log_term = 475
	last_snapshot_index = 322768509
	last_snapshot_term = 475
	latest_configuration = [{Suffrage:Voter ID:a6f23f31-d2a8-239d-401b-57d54a7521f9 Address:x.x.x.x:8300} {Suffrage:Voter ID:f6f24d3e-352e-ce6a-f5a7-1dc5ad5ad4d5 Address:10.133.0.147:8300} {Suffrage:Voter ID:3ddfa25a-fd64-ceee-ae54-f89e24c1ce70 Address:10.133.0.145:8300}]
	latest_configuration_index = 0
	num_peers = 2
	protocol_version = 3
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 475
runtime:
	arch = amd64
	cpu_count = 2
	goroutines = 4996
	max_procs = 2
	os = linux
	version = go1.16.5
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 109
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 3
	member_time = 137323
	members = 233
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 6923
	members = 12
	query_queue = 0
	query_time = 1
```
</details>

### Operating system and Environment details

Debian Buster:

```
# cat /etc/debian_version
10.10
```

Docker:

```
# docker -v
Docker version 20.10.7, build f0df350
```

Nomad:

```
# nomad -v
Nomad v1.1.2 (60638a086ef9630e2a9ba1e237e8426192a44244)
```

Envoy:

```
# envoy --version
envoy  version: 98c1c9e9a40804b93b074badad1cdf284b47d58b/1.18.3/Clean/RELEASE/BoringSSL
```

## Summary

We'll see this in a specific Envoy sidecar on a Nomad Client. For this detail, we'll say Service A (svc-a) is talking to Service B (svc-b). IPs/names are redacted here but we can provide more info securely via email.

This will start out with some connection failures:

![Screen Shot 2021-07-06 at 1 40 00 PM](https://user-images.githubusercontent.com/233334/124650948-ef1f5700-de5f-11eb-8ee9-5ca66b009085.png)

specifically, pending failure ejections:

![Screen Shot 2021-07-06 at 1 40 04 PM](https://user-images.githubusercontent.com/233334/124651003-00686380-de60-11eb-99ab-0101e60231a5.png)

Then transition to 5xx outlier ejections (expectedly):

![Screen Shot 2021-07-06 at 1 40 11 PM](https://user-images.githubusercontent.com/233334/124650982-fa728280-de5f-11eb-84cb-0c883968b078.png)

This results in all outbound connections failing with `no healthy hosts` errors.

### Statistics

From the egress sidecar, you can see the connection failures:

```
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.upstream_cx_connect_fail: 26
# ...
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.outlier_detection.ejections_active: 0
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.outlier_detection.ejections_consecutive_5xx: 5
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.outlier_detection.ejections_detected_consecutive_5xx: 5
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.outlier_detection.ejections_detected_consecutive_gateway_failure: 5
```

Eventually this will turn into active ejections, and failures. Nomad or Consul for some reason will not mark the sidecar as unhealthy, and the service will continue to serve failing traffic until restarted.

### LDS Update Failures

Looking into CDS updates, we can see that one of the clusters is unhealthy:

```
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.membership_change: 12
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.membership_degraded: 0
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.membership_excluded: 0
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.membership_healthy: 1
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.membership_total: 2
```

We have a tool we built to inspect Envoy's clusters endpoint, and (ip ranges redacted except last 2). This looks at the /clusters endpoint and aggregates our Envoy's cluster info on the `svc-b` upstream:

```
---------------------
- Cluster Instances -
---------------------
         HOST         | CX ACTIVE | CX FAILED | CX TOTAL | REQ ACTIVE | REQ TIMEOUT | REQ SUCCESS | REQ ERROR | REQ TOTAL | SUCCESS RATE | LOCAL SUCCESS RATE |    HEALTH FLAGS    | REGION | ZONE | SUBZONE | CANARY
----------------------+-----------+-----------+----------+------------+-------------+-------------+-----------+-----------+--------------+--------------------+--------------------+--------+------+---------+---------
   xx.xx.64.88:21936 |         0 |        27 |       30 |          0 |           0 |           6 |        27 |         6 |         -1.0 |               -1.0 |            healthy |        |      |         |  false
  xx.xx.64.194:21753 |         0 |         0 |        0 |          0 |           0 |           0 |         0 |         0 |         -1.0 |               -1.0 | /failed_eds_health |        |      |         |  false
```

Note here the `failed_eds_health`. However, in other sidecars with the same `svc-b` upstream:

```
---------------------
- Cluster Instances -
---------------------
         HOST         | CX ACTIVE | CX FAILED | CX TOTAL | REQ ACTIVE | REQ TIMEOUT | REQ SUCCESS | REQ ERROR | REQ TOTAL | SUCCESS RATE | LOCAL SUCCESS RATE | HEALTH FLAGS | REGION | ZONE | SUBZONE | CANARY
----------------------+-----------+-----------+----------+------------+-------------+-------------+-----------+-----------+--------------+--------------------+--------------+--------+------+---------+---------
   xx.xx.64.88:23637 |         0 |         0 |        0 |          0 |           0 |           0 |         0 |         0 |         -1.0 |               -1.0 |      healthy |        |      |         |  false
  xx.xx.64.194:23829 |         0 |         0 |        0 |          0 |           0 |           0 |         0 |         0 |         -1.0 |               -1.0 |      healthy |        |      |         |  false
```

You can see how the first Envoy sidecar is not getting updates for its cluster instances for the `svc-b` upstream.

Looking at its stats:

```
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.update_attempt: 29
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.update_empty: 0
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.update_failure: 8
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.update_no_rebuild: 0
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.update_rejected: 0
cluster.svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul.update_success: 16
# ...
listener_manager.lds.init_fetch_timeout: 0
listener_manager.lds.update_attempt: 12
listener_manager.lds.update_failure: 8
listener_manager.lds.update_rejected: 0
listener_manager.lds.update_success: 3
listener_manager.lds.update_time: 1625348489181
listener_manager.lds.version: 17241709254077376921
listener_manager.lds.version_text: ""
```

8 failures. Other sidecars have the same LDS/CDS config versions, but with correct cluster instance IPs, and no failures in updating.

### Failed Requests

We noticed this in errors:

```
[2021-07-03 12:24:42.078][1][warning][config] [source/common/config/grpc_subscription_impl.cc:119] gRPC config: initial fetch timed out for type.googleapis.com/envoy.config.endpoint.v3.ClusterLoadAssignment
[2021-07-03 21:40:39.863][1][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:101] DeltaAggregatedResources gRPC config stream closed: 13,
[2021-07-03 21:40:40.206][1][warning][config] [bazel-out/k8-opt/bin/source/common/config/_virtual_includes/grpc_stream_lib/common/config/grpc_stream.h:101] DeltaAggregatedResources gRPC config stream closed: 14, upstream connect error or disconnect/reset before headers. reset reason: connection termination
```

We also noticed this, but it seems to be related to https://github.com/hashicorp/nomad/issues/10846

```
Jul  6 16:40:13 nomad-client-xxxx consul [13260]: 2021-07-06T16:40:13.066Z [ERROR] agent.envoy: Error receiving new DeltaDiscoveryRequest; closing request channel: error="rpc error: code = Canceled desc = context canceled"
```

### Envoy Debugging

Turning on debug logging for Envoy shows this:

```
[2021-07-06 18:20:36.678][15][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1222] membership update for TLS cluster svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul added 0 removed 0
[2021-07-06 18:20:36.678][19][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1222] membership update for TLS cluster svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul added 0 removed 0
[2021-07-06 18:20:36.678][16][debug][upstream] [source/common/upstream/cluster_manager_impl.cc:1222] membership update for TLS cluster svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul added 0 removed 0
```

So that Envoy is _getting_ updates, but they're stale.

### Result: Outlier Detection Trips

Requests to stale clusters trip Envoy's outlier detection (expectedly so), and throw `upstream connect error or disconnect/reset before headers. reset reason: connection failure` errors:

```
[2021-07-06 18:22:39.680][16][debug][http] [source/common/http/conn_manager_impl.cc:261] [C35337] new stream
[2021-07-06 18:22:39.680][16][debug][http] [source/common/http/conn_manager_impl.cc:882] [C35337][S16870535376721757378] request headers complete (end_stream=true):
':authority', '0.0.0.0:8100'
':path', '/'
':method', 'POST'
'user-agent', 'curl/7.64.0'
'accept', '*/*'
'content-type', 'application/grpc'

[2021-07-06 18:22:39.680][16][debug][http] [source/common/http/filter_manager.cc:779] [C35337][S16870535376721757378] request end stream
[2021-07-06 18:22:39.680][16][debug][router] [source/common/router/router.cc:445] [C35337][S16870535376721757378] cluster 'svc-b.default.stg.internal.f73c09a4-ccb5-28bc-dc9e-1b5beff4b219.consul' match for URL '/'
[2021-07-06 18:22:39.680][16][debug][router] [source/common/router/router.cc:631] [C35337][S16870535376721757378] router decoding headers:
':authority', '0.0.0.0:8100'
':path', '/'
':method', 'POST'
':scheme', 'http'
'user-agent', 'curl/7.64.0'
'accept', '*/*'
'content-type', 'application/grpc'
'x-forwarded-proto', 'http'
'x-request-id', 'abcd1234'

[2021-07-06 18:22:39.680][16][debug][pool] [source/common/http/conn_pool_base.cc:74] queueing stream due to no available connections
[2021-07-06 18:22:39.680][16][debug][pool] [source/common/conn_pool/conn_pool_base.cc:241] trying to create new connection
[2021-07-06 18:22:39.680][16][debug][pool] [source/common/conn_pool/conn_pool_base.cc:143] creating a new connection
[2021-07-06 18:22:39.680][16][debug][client] [source/common/http/codec_client.cc:43] [C35338] connecting
[2021-07-06 18:22:39.680][16][debug][connection] [source/common/network/connection_impl.cc:860] [C35338] connecting to xx.xx.xx.88:21936
[2021-07-06 18:22:39.680][16][debug][connection] [source/common/network/connection_impl.cc:876] [C35338] connection in progress
[2021-07-06 18:22:39.680][16][debug][http2] [source/common/http/http2/codec_impl.cc:1222] [C35338] updating connection-level initial window size to 268435456
[2021-07-06 18:22:39.680][16][debug][connection] [source/common/network/connection_impl.cc:674] [C35338] delayed connection error: 111
[2021-07-06 18:22:39.680][16][debug][connection] [source/common/network/connection_impl.cc:243] [C35338] closing socket: 0
[2021-07-06 18:22:39.681][16][debug][client] [source/common/http/codec_client.cc:101] [C35338] disconnect. resetting 0 pending requests
[2021-07-06 18:22:39.681][16][debug][pool] [source/common/conn_pool/conn_pool_base.cc:393] [C35338] client disconnected, failure reason:
[2021-07-06 18:22:39.681][16][debug][router] [source/common/router/router.cc:1091] [C35337][S16870535376721757378] upstream reset: reset reason: connection failure, transport failure reason:
[2021-07-06 18:22:39.681][16][debug][http] [source/common/http/filter_manager.cc:883] [C35337][S16870535376721757378] Sending local reply with details upstream_reset_before_response_started{connection failure}
[2021-07-06 18:22:39.681][16][debug][http] [source/common/http/conn_manager_impl.cc:1469] [C35337][S16870535376721757378] encoding headers via codec (end_stream=true):
':status', '200'
'content-type', 'application/grpc'
'grpc-status', '14'
'grpc-message', 'upstream connect error or disconnect/reset before headers. reset reason: connection failure'
'date', 'Tue, 06 Jul 2021 18:22:39 GMT'
'server', 'envoy'
[2021-07-06 18:22:39.681][16][debug][connection] [source/common/network/connection_impl.cc:633] [C35337] remote close
[2021-07-06 18:22:39.681][16][debug][connection] [source/common/network/connection_impl.cc:243] [C35337] closing socket: 0
[2021-07-06 18:22:39.681][16][debug][conn_handler] [source/server/active_tcp_listener.cc:76] [C35337] adding to cleanup list
```

As described above, you can see it attempting to connect to the no-longer-existent sidecar on port `21936`, instead of the correct one at `23637`.

I deployed a new version of Consul into our environments the other day with some additional debug logging: https://github.com/bigcommerce/consul/commit/bbb90d88b331d8f9a230997e61cbf7e6ecec89d6#diff-10ee43f5dca95c57739a14e67e14d17c03f81e7f295c5c9380c48571d7470e53R649

When we had an impacted service, I went and restarted the destination service (in our case, deployed via Nomad, which caused old instances to be destroyed and new to come up with new port assignments, etc), and the logging basically gave me this:

```
Invoking all xDS resource handlers and sending changed data if there are any: service_id=_nomad-task-63f85afa-27a9-aad8-76f0-9742022aab84-group-source-service-sidecar-proxy xdsVersion=v3
createDeltaResponse: resourceVersions loop: service_id=_nomad-task-63f85afa-27a9-aad8-76f0-9742022aab84-group-source-service-sidecar-proxy typeUrl=type.googleapis.com/envoy.config.listener.v3.Listener xdsVersion=v3 name=:destination-service127.0.0.1:8103 envoyVers=d11c05d06147c85870ca3316b1d807a0236d38e9dc580351bdf7cb7dc6ff38c6
..
createDeltaResponse: no relevant updates and sentToEnvoyOnce: service_id=_nomad-task-63f85afa-27a9-aad8-76f0-9742022aab84-group-source-service-sidecar-proxy typeUrl=type.googleapis.com/envoy.config.listener.v3.Listener xdsVersion=v3
SendIfNew: no response to send: service_id=_nomad-task-63f85afa-27a9-aad8-76f0-9742022aab84-group-source-service-sidecar-proxy typeUrl=type.googleapis.com/envoy.config.listener.v3.Listener xdsVersion=v3
```

Based on the above, it looks like Consul is determining there are no changes to be propagated down to Envoy - either it thinks they were already sent, it's stuck serving up some cached copy of the upstreams for a service (I call this out because I remember seeing something with respect to proxycfg caching), or there's something with the delta comparison that's not quite right.

Effectively, based on the logging here: https://github.com/bigcommerce/consul/commit/bbb90d88b331d8f9a230997e61cbf7e6ecec89d6#diff-10ee43f5dca95c57739a14e67e14d17c03f81e7f295c5c9380c48571d7470e53R649, `t.resourceVersions` I believe, does not include a differing hash/version for the destination service.

I've got a new build which is WIP that adds some additional trace logging earlier on in the process - the determination of upstream members for a service, which I hope to ship over the weekend.

---

As another anecdote, we had a further + slightly different occurence of this last night, which I think is related. After we dropped a `service-splitter` to split between an existing service and a new service, some (not all) of our Consul Connect/Envoy instances went crazy with the following:

```
[ERROR] agent.envoy.xds: got error response from envoy proxy:
  service_id=_nomad-task-deabbb0e-4737-479f-312e-83ff6107360f-group-source-service-sidecar-proxy typeUrl=type.googleapis.com/envoy.config.route.v3.RouteConfiguration xdsVersion=v3 nonce=000434bd error="rpc error: code = Internal desc = route: unknown weighted cluster 'new-service-to-split-to.default.consul-dc.e6a3564e-cdc3-2dd2-87c3-5d8a6dd63351.consul'"
```

It seems as if in this situation an update to CDS to deliver the `new-service-to-split-to` cluster was skipped or missed, or for some other reason didn't land in Envoy properly. We've got a couple of occurrences of this still happening, which hopefully I can pull some trace level logs from shortly.

--- a/.changelog/10987.txt
+++ b/.changelog/10987.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+xds: fixed a bug where Envoy sidecars could enter a state where they failed to receive xds updates from Consul
+```

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -189,6 +189,10 @@ func (s *Server) processDelta(stream ADSDeltaStream, reqCh <-chan *envoy_discove
 			// index and hash the xDS structures
 			newResourceMap := indexResources(generator.Logger, newRes)
 
+			if s.ResourceMapMutateFn != nil {
+				s.ResourceMapMutateFn(newResourceMap)
+			}
+
 			if err := populateChildIndexMap(newResourceMap); err != nil {
 				return status.Errorf(codes.Unavailable, "failed to index xDS resource versions: %v", err)
 			}

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -568,16 +568,10 @@ func (t *xDSDeltaType) ack(nonce string) {
 
 	for name, obj := range pending {
 		if obj.Remove {
-			if obj.Version != "" {
-				panic("ACK: version is not empty for: " + name)
-			}
 			delete(t.resourceVersions, name)
 			continue
 		}
 
-		if obj.Version == "" {
-			panic("ACK: version is empty for: " + name)
-		}
 		t.resourceVersions[name] = obj.Version
 		if t.childType != nil {
 			// This branch only matters on UPDATE, since we already have
@@ -751,17 +745,11 @@ func (t *xDSDeltaType) createDeltaResponse(
 	realUpdates := make(map[string]PendingUpdate)
 	for name, obj := range updates {
 		if obj.Remove {
-			if obj.Version != "" {
-				panic("DIFF: version is not empty for: " + name)
-			}
 			if remove {
 				resp.RemovedResources = append(resp.RemovedResources, name)
 				realUpdates[name] = PendingUpdate{Remove: true}
 			}
 		} else if upsert {
-			if obj.Version == "" {
-				panic("DIFF: version is empty for: " + name)
-			}
 			resources, ok := resourceMap.Index[t.typeURL]
 			if !ok {
 				return nil, nil, fmt.Errorf("unknown type url: %s", t.typeURL)

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -5,6 +5,7 @@ import (
 	"encoding/hex"
 	"fmt"
 	"sort"
+	"strings"
 	"sync/atomic"
 	"time"
 
@@ -69,6 +70,10 @@ const (
 )
 
 func (s *Server) processDelta(stream ADSDeltaStream, reqCh <-chan *envoy_discovery_v3.DeltaDiscoveryRequest) error {
+	slowHackValidUntil := time.Now().Add(1 * time.Minute)
+	slowHackDisabled := false
+	s.Logger.Info("HACK: forcing endpoints containing the string 'slow' to wait", "until", slowHackValidUntil)
+
 	// Loop state
 	var (
 		cfgSnap     *proxycfg.ConfigSnapshot
@@ -188,6 +193,26 @@ func (s *Server) processDelta(stream ADSDeltaStream, reqCh <-chan *envoy_discove
 
 			// index and hash the xDS structures
 			newResourceMap := indexResources(generator.Logger, newRes)
+
+			// HACK to prove a theory
+			if !slowHackDisabled && time.Now().Before(slowHackValidUntil) {
+				if em, ok := newResourceMap.Index[EndpointType]; ok {
+					var deleteme []string
+					for k, _ := range em {
+						if strings.Contains(k, "slow") {
+							deleteme = append(deleteme, k)
+						}
+					}
+					for _, k := range deleteme {
+						delete(em, k)
+					}
+				}
+			} else {
+				if !slowHackDisabled {
+					slowHackDisabled = true
+					s.Logger.Info("HACK: forcing endpoints containing the string 'slow' to no longer wait")
+				}
+			}
 
 			if err := populateChildIndexMap(newResourceMap); err != nil {
 				return status.Errorf(codes.Unavailable, "failed to index xDS resource versions: %v", err)

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -710,18 +710,13 @@ func (t *xDSDeltaType) createDeltaResponse(
 		// Walk the list of things currently stored in envoy
 		for name, envoyVers := range t.resourceVersions {
 			if t.subscribed(name) {
-				currVers, ok := currentVersions[name]
-				if ok {
+				if currVers, ok := currentVersions[name]; ok {
 					if currVers != envoyVers {
 						if upsert {
 							hasRelevantUpdates = true
 						}
 						updates[name] = PendingUpdate{Version: currVers}
 					}
-				} else {
-					// Should we do anything here? This might be an eventual
-					// consistency issue, but it's unclear what we could do
-					// other than just wait until it resolves.
 				}
 			}
 		}
@@ -731,16 +726,11 @@ func (t *xDSDeltaType) createDeltaResponse(
 			if _, known := t.resourceVersions[name]; known {
 				continue
 			}
-			currVers, ok := currentVersions[name]
-			if ok {
+			if currVers, ok := currentVersions[name]; ok {
 				updates[name] = PendingUpdate{Version: currVers}
 				if upsert {
 					hasRelevantUpdates = true
 				}
-			} else {
-				// Should we do anything here? This might be an eventual
-				// consistency issue, but it's unclear what we could do
-				// other than just wait until it resolves.
 			}
 		}
 	}

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -384,6 +384,10 @@ type xDSDeltaType struct {
 	// sentToEnvoyOnce is true after we've sent one response to envoy.
 	sentToEnvoyOnce bool
 
+	// subscriptions is the list of currently subscribed envoy resources. If
+	// the wildcard is in play this will be empty.
+	subscriptions map[string]struct{}
+
 	// resourceVersions is the current view of CONFIRMED/ACKed updates to
 	// envoy's view of the loaded resources.
 	//
@@ -398,7 +402,16 @@ type xDSDeltaType struct {
 	pendingUpdates map[string]map[string]PendingUpdate
 }
 
+func (t *xDSDeltaType) subscribed(name string) bool {
+	if t.wildcard {
+		return true
+	}
+	_, subscribed := t.subscriptions[name]
+	return subscribed
+}
+
 type PendingUpdate struct {
+	Remove         bool
 	Version        string
 	ChildResources []string // optional
 }
@@ -414,6 +427,7 @@ func newDeltaType(
 		stream:           stream,
 		typeURL:          typeUrl,
 		allowEmptyFn:     allowEmptyFn,
+		subscriptions:    make(map[string]struct{}),
 		resourceVersions: make(map[string]string),
 		pendingUpdates:   make(map[string]map[string]PendingUpdate),
 	}
@@ -484,6 +498,11 @@ func (t *xDSDeltaType) Recv(req *envoy_discovery_v3.DeltaDiscoveryRequest) bool 
 		logger.Trace("setting initial resource versions for stream",
 			"resources", req.InitialResourceVersions)
 		t.resourceVersions = req.InitialResourceVersions
+		if !t.wildcard {
+			for k, _ := range req.InitialResourceVersions {
+				t.subscriptions[k] = struct{}{}
+			}
+		}
 	}
 
 	if !t.wildcard {
@@ -509,8 +528,13 @@ func (t *xDSDeltaType) Recv(req *envoy_discovery_v3.DeltaDiscoveryRequest) bool 
 			//
 			// We handle that here by ALWAYS wiping the version so the diff
 			// decides to send the value.
-			_, alreadySubscribed := t.resourceVersions[name]
-			t.resourceVersions[name] = "" // start with no version
+			_, alreadySubscribed := t.subscriptions[name]
+			t.subscriptions[name] = struct{}{}
+
+			// Reset the tracked version so we force a reply.
+			if _, alreadyTracked := t.resourceVersions[name]; alreadyTracked {
+				t.resourceVersions[name] = ""
+			}
 
 			if alreadySubscribed {
 				logger.Trace("re-subscribing resource for stream", "resource", name)
@@ -520,11 +544,12 @@ func (t *xDSDeltaType) Recv(req *envoy_discovery_v3.DeltaDiscoveryRequest) bool 
 		}
 
 		for _, name := range req.ResourceNamesUnsubscribe {
-			if _, ok := t.resourceVersions[name]; !ok {
+			if _, ok := t.subscriptions[name]; !ok {
 				continue
 			}
-			delete(t.resourceVersions, name)
+			delete(t.subscriptions, name)
 			logger.Trace("unsubscribing resource for stream", "resource", name)
+			// NOTE: we'll let the normal differential comparison handle cleaning up resourceVersions
 		}
 	}
 
@@ -538,26 +563,37 @@ func (t *xDSDeltaType) ack(nonce string) {
 	}
 
 	for name, obj := range pending {
-		if obj.Version == "" {
+		if obj.Remove {
+			if obj.Version != "" {
+				panic("ACK: version is not empty for: " + name)
+			}
 			delete(t.resourceVersions, name)
-		} else {
-			t.resourceVersions[name] = obj.Version
+			continue
 		}
-		if t.childType != nil && obj.Version != "" {
+
+		if obj.Version == "" {
+			panic("ACK: version is empty for: " + name)
+		}
+		t.resourceVersions[name] = obj.Version
+		if t.childType != nil {
 			// This branch only matters on UPDATE, since we already have
 			// mechanisms to clean up orphaned resources.
 			for _, childName := range obj.ChildResources {
-				if _, exist := t.childType.resourceVersions[childName]; exist {
-					t.generator.Logger.Trace(
-						"triggering implicit update of resource",
-						"typeUrl", t.typeURL,
-						"resource", name,
-						"childTypeUrl", t.childType.typeURL,
-						"childResource", childName,
-					)
-					// Basically manifest this as a re-subscribe
-					t.childType.resourceVersions[childName] = ""
+				if _, exist := t.childType.resourceVersions[childName]; !exist {
+					continue
 				}
+				if !t.subscribed(childName) {
+					continue
+				}
+				t.generator.Logger.Trace(
+					"triggering implicit update of resource",
+					"typeUrl", t.typeURL,
+					"resource", name,
+					"childTypeUrl", t.childType.typeURL,
+					"childResource", childName,
+				)
+				// Basically manifest this as a re-subscribe/re-sync
+				t.childType.resourceVersions[childName] = ""
 			}
 		}
 	}
@@ -640,30 +676,73 @@ func (t *xDSDeltaType) createDeltaResponse(
 		hasRelevantUpdates = false
 		updates            = make(map[string]PendingUpdate)
 	)
-	// First find things that need updating or deleting
-	for name, envoyVers := range t.resourceVersions {
-		currVers, ok := currentVersions[name]
-		if !ok {
-			if remove {
-				hasRelevantUpdates = true
+
+	if t.wildcard {
+		// First find things that need updating or deleting
+		for name, envoyVers := range t.resourceVersions {
+			currVers, ok := currentVersions[name]
+			if !ok {
+				if remove {
+					hasRelevantUpdates = true
+				}
+				updates[name] = PendingUpdate{Remove: true}
+			} else if currVers != envoyVers {
+				if upsert {
+					hasRelevantUpdates = true
+				}
+				updates[name] = PendingUpdate{Version: currVers}
 			}
-			updates[name] = PendingUpdate{Version: ""}
-		} else if currVers != envoyVers {
+		}
+
+		// Now find new things
+		for name, currVers := range currentVersions {
+			if _, known := t.resourceVersions[name]; known {
+				continue
+			}
 			if upsert {
 				hasRelevantUpdates = true
 			}
 			updates[name] = PendingUpdate{Version: currVers}
 		}
-	}
+	} else {
+		// First find things that need updating or deleting
 
-	// Now find new things
-	if t.wildcard {
-		for name, currVers := range currentVersions {
-			if _, ok := t.resourceVersions[name]; !ok {
+		// Walk the list of things currently stored in envoy
+		for name, envoyVers := range t.resourceVersions {
+			if t.subscribed(name) {
+				currVers, ok := currentVersions[name]
+				if ok {
+					if currVers != envoyVers {
+						if upsert {
+							hasRelevantUpdates = true
+						}
+						updates[name] = PendingUpdate{Version: currVers}
+					}
+				} else {
+					// TODO: should we do anything; this might be an eventual consistency hole
+					// i guess we'll let it ride?
+				}
+			} else {
+				if remove {
+					hasRelevantUpdates = true
+				}
+				updates[name] = PendingUpdate{Remove: true}
+			}
+		}
+
+		// Now find new things not in envoy yet
+		for name, _ := range t.subscriptions {
+			if _, known := t.resourceVersions[name]; known {
+				continue
+			}
+			currVers, ok := currentVersions[name]
+			if ok {
 				updates[name] = PendingUpdate{Version: currVers}
 				if upsert {
 					hasRelevantUpdates = true
 				}
+			} else {
+				// TODO: should we do anything here? or should we just wait and see?
 			}
 		}
 	}
@@ -679,12 +758,18 @@ func (t *xDSDeltaType) createDeltaResponse(
 	}
 	realUpdates := make(map[string]PendingUpdate)
 	for name, obj := range updates {
-		if obj.Version == "" {
+		if obj.Remove {
+			if obj.Version != "" {
+				panic("DIFF: version is not empty for: " + name)
+			}
 			if remove {
 				resp.RemovedResources = append(resp.RemovedResources, name)
-				realUpdates[name] = PendingUpdate{Version: ""}
+				realUpdates[name] = PendingUpdate{Remove: true}
 			}
 		} else if upsert {
+			if obj.Version == "" {
+				panic("DIFF: version is empty for: " + name)
+			}
 			resources, ok := resourceMap.Index[t.typeURL]
 			if !ok {
 				return nil, nil, fmt.Errorf("unknown type url: %s", t.typeURL)

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -384,7 +384,7 @@ type xDSDeltaType struct {
 	// sentToEnvoyOnce is true after we've sent one response to envoy.
 	sentToEnvoyOnce bool
 
-	// subscriptions is the set of currently subscribed envoy resources. 
+	// subscriptions is the set of currently subscribed envoy resources.
 	// If wildcard == true, this will be empty.
 	subscriptions map[string]struct{}
 

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -499,7 +499,7 @@ func (t *xDSDeltaType) Recv(req *envoy_discovery_v3.DeltaDiscoveryRequest) bool 
 			"resources", req.InitialResourceVersions)
 		t.resourceVersions = req.InitialResourceVersions
 		if !t.wildcard {
-			for k, _ := range req.InitialResourceVersions {
+			for k := range req.InitialResourceVersions {
 				t.subscriptions[k] = struct{}{}
 			}
 		}
@@ -727,7 +727,7 @@ func (t *xDSDeltaType) createDeltaResponse(
 		}
 
 		// Now find new things not in envoy yet
-		for name, _ := range t.subscriptions {
+		for name := range t.subscriptions {
 			if _, known := t.resourceVersions[name]; known {
 				continue
 			}

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -384,8 +384,8 @@ type xDSDeltaType struct {
 	// sentToEnvoyOnce is true after we've sent one response to envoy.
 	sentToEnvoyOnce bool
 
-	// subscriptions is the list of currently subscribed envoy resources. If
-	// the wildcard is in play this will be empty.
+	// subscriptions is the set of currently subscribed envoy resources. 
+	// If wildcard == true, this will be empty.
 	subscriptions map[string]struct{}
 
 	// resourceVersions is the current view of CONFIRMED/ACKed updates to

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -5,7 +5,6 @@ import (
 	"encoding/hex"
 	"fmt"
 	"sort"
-	"strings"
 	"sync/atomic"
 	"time"
 
@@ -70,10 +69,6 @@ const (
 )
 
 func (s *Server) processDelta(stream ADSDeltaStream, reqCh <-chan *envoy_discovery_v3.DeltaDiscoveryRequest) error {
-	slowHackValidUntil := time.Now().Add(1 * time.Minute)
-	slowHackDisabled := false
-	s.Logger.Info("HACK: forcing endpoints containing the string 'slow' to wait", "until", slowHackValidUntil)
-
 	// Loop state
 	var (
 		cfgSnap     *proxycfg.ConfigSnapshot
@@ -193,26 +188,6 @@ func (s *Server) processDelta(stream ADSDeltaStream, reqCh <-chan *envoy_discove
 
 			// index and hash the xDS structures
 			newResourceMap := indexResources(generator.Logger, newRes)
-
-			// HACK to prove a theory
-			if !slowHackDisabled && time.Now().Before(slowHackValidUntil) {
-				if em, ok := newResourceMap.Index[EndpointType]; ok {
-					var deleteme []string
-					for k, _ := range em {
-						if strings.Contains(k, "slow") {
-							deleteme = append(deleteme, k)
-						}
-					}
-					for _, k := range deleteme {
-						delete(em, k)
-					}
-				}
-			} else {
-				if !slowHackDisabled {
-					slowHackDisabled = true
-					s.Logger.Info("HACK: forcing endpoints containing the string 'slow' to no longer wait")
-				}
-			}
 
 			if err := populateChildIndexMap(newResourceMap); err != nil {
 				return status.Errorf(codes.Unavailable, "failed to index xDS resource versions: %v", err)

--- a/agent/xds/delta.go
+++ b/agent/xds/delta.go
@@ -719,14 +719,10 @@ func (t *xDSDeltaType) createDeltaResponse(
 						updates[name] = PendingUpdate{Version: currVers}
 					}
 				} else {
-					// TODO: should we do anything; this might be an eventual consistency hole
-					// i guess we'll let it ride?
+					// Should we do anything here? This might be an eventual
+					// consistency issue, but it's unclear what we could do
+					// other than just wait until it resolves.
 				}
-			} else {
-				if remove {
-					hasRelevantUpdates = true
-				}
-				updates[name] = PendingUpdate{Remove: true}
 			}
 		}
 
@@ -742,7 +738,9 @@ func (t *xDSDeltaType) createDeltaResponse(
 					hasRelevantUpdates = true
 				}
 			} else {
-				// TODO: should we do anything here? or should we just wait and see?
+				// Should we do anything here? This might be an eventual
+				// consistency issue, but it's unclear what we could do
+				// other than just wait until it resolves.
 			}
 		}
 	}

--- a/agent/xds/delta_test.go
+++ b/agent/xds/delta_test.go
@@ -430,7 +430,7 @@ func TestServer_DeltaAggregatedResources_v3_SlowEndpointPopulation(t *testing.T)
 		}
 		if em, ok := resourceMap.Index[EndpointType]; ok {
 			var deleteme []string
-			for k, _ := range em {
+			for k := range em {
 				if strings.Contains(k, "geo-cache") {
 					deleteme = append(deleteme, k)
 				}

--- a/agent/xds/delta_test.go
+++ b/agent/xds/delta_test.go
@@ -2,6 +2,7 @@ package xds
 
 import (
 	"errors"
+	"strings"
 	"sync/atomic"
 	"testing"
 	"time"
@@ -399,6 +400,153 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_HTTP2(t *testing.T) {
 		envoy.SendDeltaReqACK(t, RouteType, 5)
 
 		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+	})
+
+	envoy.Close()
+	select {
+	case err := <-errCh:
+		require.NoError(t, err)
+	case <-time.After(50 * time.Millisecond):
+		t.Fatalf("timed out waiting for handler to finish")
+	}
+}
+
+func TestServer_DeltaAggregatedResources_v3_SlowEndpointPopulation(t *testing.T) {
+	// This illustrates a scenario related to https://github.com/hashicorp/consul/issues/10563
+
+	aclResolve := func(id string) (acl.Authorizer, error) {
+		// Allow all
+		return acl.RootAuthorizer("manage"), nil
+	}
+	scenario := newTestServerDeltaScenario(t, aclResolve, "web-sidecar-proxy", "", 0)
+	server, mgr, errCh, envoy := scenario.server, scenario.mgr, scenario.errCh, scenario.envoy
+
+	// This mutateFn causes any endpoint with a name containing "geo-cache" to be
+	// omitted from the response while the hack is active.
+	var slowHackDisabled uint32
+	server.ResourceMapMutateFn = func(resourceMap *IndexedResources) {
+		if atomic.LoadUint32(&slowHackDisabled) == 1 {
+			return
+		}
+		if em, ok := resourceMap.Index[EndpointType]; ok {
+			var deleteme []string
+			for k, _ := range em {
+				if strings.Contains(k, "geo-cache") {
+					deleteme = append(deleteme, k)
+				}
+			}
+			for _, k := range deleteme {
+				delete(em, k)
+			}
+		}
+	}
+
+	sid := structs.NewServiceID("web-sidecar-proxy", nil)
+
+	// Register the proxy to create state needed to Watch() on
+	mgr.RegisterProxy(t, sid)
+
+	var snap *proxycfg.ConfigSnapshot
+	runStep(t, "get into initial state", func(t *testing.T) {
+		snap = newTestSnapshot(t, nil, "")
+
+		// Send initial cluster discover.
+		envoy.SendDeltaReq(t, ClusterType, &envoy_discovery_v3.DeltaDiscoveryRequest{})
+
+		// Check no response sent yet
+		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+		requireProtocolVersionGauge(t, scenario, "v3", 1)
+
+		// Deliver a new snapshot (tcp with one tcp upstream)
+		mgr.DeliverConfig(t, sid, snap)
+
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: ClusterType,
+			Nonce:   hexString(1),
+			Resources: makeTestResources(t,
+				makeTestCluster(t, snap, "tcp:local_app"),
+				makeTestCluster(t, snap, "tcp:db"),
+				makeTestCluster(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		// Envoy then tries to discover endpoints for those clusters.
+		envoy.SendDeltaReq(t, EndpointType, &envoy_discovery_v3.DeltaDiscoveryRequest{
+			ResourceNamesSubscribe: []string{
+				"db.default.dc1.internal.11111111-2222-3333-4444-555555555555.consul",
+				"geo-cache.default.dc1.query.11111111-2222-3333-4444-555555555555.consul",
+			},
+		})
+
+		// It also (in parallel) issues the cluster ACK
+		envoy.SendDeltaReqACK(t, ClusterType, 1)
+
+		// We should get a response immediately since the config is already present in
+		// the server for endpoints. Note that this should not be racy if the server
+		// is behaving well since the Cluster send above should be blocked until we
+		// deliver a new config version.
+		//
+		// NOTE: we do NOT return back geo-cache yet
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: EndpointType,
+			Nonce:   hexString(2),
+			Resources: makeTestResources(t,
+				makeTestEndpoints(t, snap, "tcp:db"),
+				// makeTestEndpoints(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		// And no other response yet
+		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+		// Envoy now sends listener request
+		envoy.SendDeltaReq(t, ListenerType, nil)
+
+		// It also (in parallel) issues the endpoint ACK
+		envoy.SendDeltaReqACK(t, EndpointType, 2)
+
+		// And should get a response immediately.
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: ListenerType,
+			Nonce:   hexString(3),
+			Resources: makeTestResources(t,
+				makeTestListener(t, snap, "tcp:public_listener"),
+				makeTestListener(t, snap, "tcp:db"),
+				makeTestListener(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		// And no other response yet
+		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+		// ACKs the listener
+		envoy.SendDeltaReqACK(t, ListenerType, 3)
+	})
+
+	// Disable hack. Need to wait for one more event to wake up the loop.
+	atomic.StoreUint32(&slowHackDisabled, 1)
+
+	runStep(t, "delayed endpoint update finally comes in", func(t *testing.T) {
+		// Trigger the xds.Server select{} to wake up and notice our hack is disabled.
+		// The actual contents of this change are irrelevant.
+		snap = newTestSnapshot(t, snap, "")
+		mgr.DeliverConfig(t, sid, snap)
+
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: EndpointType,
+			Nonce:   hexString(4),
+			Resources: makeTestResources(t,
+				makeTestEndpoints(t, snap, "tcp:geo-cache"),
+			),
+		})
+
+		// And no other response yet
+		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
+
+		// It also (in parallel) issues the endpoint ACK
+		envoy.SendDeltaReqACK(t, EndpointType, 4)
+
 	})
 
 	envoy.Close()

--- a/agent/xds/delta_test.go
+++ b/agent/xds/delta_test.go
@@ -100,6 +100,7 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_TCP(t *testing.T) {
 			Resources: makeTestResources(t,
 				makeTestEndpoints(t, snap, "tcp:db"),
 				// SAME_AS_INITIAL_VERSION: makeTestEndpoints(t, snap, "tcp:geo-cache"),
+				// SAME_AS_INITIAL_VERSION: "fake-endpoints",
 			),
 		})
 
@@ -123,23 +124,11 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_TCP(t *testing.T) {
 			),
 		})
 
-		// cleanup unused resources now that we've created/updated relevant things
-		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
-			TypeUrl: EndpointType,
-			Nonce:   hexString(4),
-			RemovedResources: []string{
-				"fake-endpoints", // correcting the errant subscription
-			},
-		})
-
 		// And no other response yet
 		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
 
 		// ACKs the listener
 		envoy.SendDeltaReqACK(t, ListenerType, 3)
-
-		// ACK the endpoint removal
-		envoy.SendDeltaReqACK(t, EndpointType, 4)
 
 		// If we re-subscribe to something even if there are no changes we get a
 		// fresh copy.
@@ -151,13 +140,13 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_TCP(t *testing.T) {
 
 		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
 			TypeUrl: EndpointType,
-			Nonce:   hexString(5),
+			Nonce:   hexString(4),
 			Resources: makeTestResources(t,
 				makeTestEndpoints(t, snap, "tcp:geo-cache"),
 			),
 		})
 
-		envoy.SendDeltaReqACK(t, EndpointType, 5)
+		envoy.SendDeltaReqACK(t, EndpointType, 4)
 
 		// And no other response yet
 		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
@@ -202,13 +191,13 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_TCP(t *testing.T) {
 		})
 		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
 			TypeUrl: EndpointType,
-			Nonce:   hexString(6),
+			Nonce:   hexString(5),
 			Resources: makeTestResources(t,
 				makeTestEndpoints(t, snap, "tcp:db"),
 			),
 		})
 
-		envoy.SendDeltaReqACK(t, EndpointType, 6)
+		envoy.SendDeltaReqACK(t, EndpointType, 5)
 
 		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
 	})
@@ -222,24 +211,24 @@ func TestServer_DeltaAggregatedResources_v3_BasicProtocol_TCP(t *testing.T) {
 		// Send envoy an EDS update.
 		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
 			TypeUrl: EndpointType,
+			Nonce:   hexString(6),
+			Resources: makeTestResources(t,
+				makeTestEndpoints(t, snap, "tcp:db[0]"),
+			),
+		})
+
+		envoy.SendDeltaReqNACK(t, EndpointType, 6, &rpcstatus.Status{})
+
+		// Send it again.
+		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
+			TypeUrl: EndpointType,
 			Nonce:   hexString(7),
 			Resources: makeTestResources(t,
 				makeTestEndpoints(t, snap, "tcp:db[0]"),
 			),
 		})
 
-		envoy.SendDeltaReqNACK(t, EndpointType, 7, &rpcstatus.Status{})
-
-		// Send it again.
-		assertDeltaResponseSent(t, envoy.deltaStream.sendCh, &envoy_discovery_v3.DeltaDiscoveryResponse{
-			TypeUrl: EndpointType,
-			Nonce:   hexString(8),
-			Resources: makeTestResources(t,
-				makeTestEndpoints(t, snap, "tcp:db[0]"),
-			),
-		})
-
-		envoy.SendDeltaReqACK(t, EndpointType, 8)
+		envoy.SendDeltaReqACK(t, EndpointType, 7)
 
 		assertDeltaChanBlocked(t, envoy.deltaStream.sendCh)
 	})

--- a/agent/xds/delta_test.go
+++ b/agent/xds/delta_test.go
@@ -429,14 +429,10 @@ func TestServer_DeltaAggregatedResources_v3_SlowEndpointPopulation(t *testing.T)
 			return
 		}
 		if em, ok := resourceMap.Index[EndpointType]; ok {
-			var deleteme []string
 			for k := range em {
 				if strings.Contains(k, "geo-cache") {
-					deleteme = append(deleteme, k)
+					delete(em, k)
 				}
-			}
-			for _, k := range deleteme {
-				delete(em, k)
 			}
 		}
 	}

--- a/agent/xds/server.go
+++ b/agent/xds/server.go
@@ -152,6 +152,9 @@ type Server struct {
 
 	DisableV2Protocol bool
 
+	// ResourceMapMutateFn exclusively exists for testing purposes.
+	ResourceMapMutateFn func(resourceMap *IndexedResources)
+
 	activeStreams *activeStreamCounters
 }
 

--- a/agent/xds/xds_protocol_helpers_test.go
+++ b/agent/xds/xds_protocol_helpers_test.go
@@ -177,7 +177,9 @@ func newTestServerScenarioInner(
 		nil, /*checkFetcher HTTPCheckFetcher*/
 		nil, /*cfgFetcher ConfigFetcher*/
 	)
-	s.AuthCheckFrequency = authCheckFrequency
+	if authCheckFrequency > 0 {
+		s.AuthCheckFrequency = authCheckFrequency
+	}
 
 	errCh := make(chan error, 1)
 	go func() {

